esd_services_api_client/nexus/input/payload_reader.py

```diff
@@ -0,0 +1,75 @@
+"""
+ Code infrastructure for manipulating payload received from Crystal SAS URI
+"""
+
+#  Copyright (c) 2023. ECCO Sneaks & Data
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#
+
+from dataclasses import dataclass
+
+from typing import final, Optional, Type
+
+from adapta.storage.models.format import DictJsonSerializationFormat
+from adapta.utils import session_with_retries
+
+from dataclasses_json import DataClassJsonMixin
+
+
+@dataclass
+class AlgorithmPayload(DataClassJsonMixin):
+    """
+    Base class for algorithm payload
+    """
+
+
+@final
+class AlgorithmPayloadReader:
+    """
+    Crystal Payload Reader - receives the payload from the URI and deserializes it into the specified type
+    """
+
+    async def __aenter__(self):
+        if not self._http:
+            self._http = session_with_retries()
+        http_response = self._http.get(url=self._payload_uri)
+        http_response.raise_for_status()
+        self._payload = self._payload_type.from_dict(
+            DictJsonSerializationFormat().deserialize(http_response.content)
+        )
+        return self
+
+    async def __aexit__(self, exc_type, exc_val, exc_tb):
+        self._http.close()
+        self._http = None
+
+    def __init__(self, payload_uri: str, payload_type: Type[AlgorithmPayload]):
+        self._http = session_with_retries()
+        self._payload: Optional[AlgorithmPayload] = None
+        self._payload_uri = payload_uri
+        self._payload_type = payload_type
+
+    @property
+    def payload_uri(self) -> Optional[str]:
```

Comment:
The return type here should not be optional, because `self._payload_uri` is initialized from constructor argument `payload_uri` which is not optional `str`.